File: Sources/Chunk/BedrockChunkSubChunkAccess.swift
import Foundation

enum BedrockSubChunkBacking: Equatable {
    case subChunk(key: Data)
    case legacyTerrain(key: Data)
}

struct BedrockStoredSubChunk {
    let yIndex: Int8
    let subChunk: BedrockSubChunk
    let backing: BedrockSubChunkBacking
}

/// Unified access to independent SubChunkPrefix records, their legacy 0x34
/// second block layer, and pre-0.17 LegacyTerrain records. The rest of the
/// editor therefore sees a coordinate-oriented block model while persistence
/// remains faithful to each historical format.
enum BedrockChunkSubChunkAccess {
    static func records(database: MojangLevelDB, position: ChunkPosition) throws -> [BedrockStoredSubChunk] {
        var prefix = Data()
        prefix.appendLE(position.x)
        prefix.appendLE(position.z)
        let entries = try database.entries(prefix: prefix, includeValues: true, limit: 0)

        var output = [BedrockStoredSubChunk]()
        var occupied = Set<Int8>()
        var legacyTerrainRecord: (key: Data, value: Data)?
        var extraData: BedrockLegacyBlockExtraData?
        var observedNumericVersion: UInt8?

        for entry in entries {
            guard let parsed = BedrockDBKey.parse(entry.key), parsed.position == position, let raw = entry.value else { continue }
            if parsed.recordType == .subChunk, let keyY = parsed.subChunkIndex {
                let decoded = try BedrockSubChunk.decode(raw, keyYIndex: keyY)
                let logicalY = decoded.yIndex ?? keyY
                if decoded.isLegacyNumeric {
                    observedNumericVersion = observedNumericVersion ?? decoded.version
                }
                output.append(BedrockStoredSubChunk(yIndex: logicalY, subChunk: decoded, backing: .subChunk(key: entry.key)))
                occupied.insert(logicalY)
            } else if parsed.recordType == .legacyTerrain {
                legacyTerrainRecord = (entry.key, raw)
            } else if parsed.recordType == .legacyBlockExtraData {
                extraData = try BedrockLegacyBlockExtraData.decode(raw)
            }
        }

        // Attach 0x34 as virtual layer 1 only to v0/v2...v7 0x2F records.
        // LegacyTerrain is a separate pre-Anvil format and must not inherit the
        // SubChunk-specific 0x34 coordinate interpretation.
        if let extraData {
            output = output.map { record in
                guard record.subChunk.isLegacyNumeric,
                      case .subChunk = record.backing,
                      let layer1 = extraData.storage(subChunkY: record.yIndex) else { return record }
                return BedrockStoredSubChunk(
                    yIndex: record.yIndex,
                    subChunk: record.subChunk.withLegacyExtraLayer(layer1),
                    backing: record.backing
                )
            }

            // An all-air primary SubChunk can be absent while 0x34 still has
            // extra blocks. Expose such slices instead of silently hiding the
            // second layer. Materialising/editing them later creates a numeric
            // 0x2F layer-0 record using the observed legacy version (or v0).
            let fallbackVersion = observedNumericVersion ?? 0
            for y in extraData.subChunkYIndices.sorted() where !occupied.contains(y) {
                guard (0...15).contains(Int(y)), let layer1 = extraData.storage(subChunkY: y) else { continue }
                var base = try BedrockSubChunk.emptyLegacy(version: fallbackVersion, yIndex: y)
                base = base.withLegacyExtraLayer(layer1)
                let key = BedrockDBKey.subChunk(
                    x: position.x, z: position.z, dimension: position.dimension, index: y
                )
                output.append(BedrockStoredSubChunk(yIndex: y, subChunk: base, backing: .subChunk(key: key)))
                occupied.insert(y)
            }
        }

        if let record = legacyTerrainRecord {
            let terrain = try BedrockLegacyTerrain.decode(record.value)
            for rawY in 0..<8 {
                let y = Int8(rawY)
                guard !occupied.contains(y) else { continue }
                output.append(BedrockStoredSubChunk(
                    yIndex: y,
                    subChunk: try terrain.subChunk(yIndex: y),
                    backing: .legacyTerrain(key: record.key)
                ))
            }
        }
        return output.sorted { lhs, rhs in
            if lhs.yIndex != rhs.yIndex { return lhs.yIndex < rhs.yIndex }
            switch (lhs.backing, rhs.backing) {
            case (.subChunk, .legacyTerrain): return true
            case (.legacyTerrain, .subChunk): return false
            default: return false
            }
        }
    }

    static func record(database: MojangLevelDB, position: ChunkPosition, yIndex: Int8) throws -> BedrockStoredSubChunk? {
        // Fast path for the overwhelmingly common key-Y == logical-Y case.
        let key = BedrockDBKey.subChunk(
            x: position.x, z: position.z, dimension: position.dimension, index: yIndex
        )
        if let raw = try database.get(key) {
            var decoded = try BedrockSubChunk.decode(raw, keyYIndex: yIndex)
            let logicalY = decoded.yIndex ?? yIndex
            if logicalY == yIndex {
                if decoded.isLegacyNumeric,
                   let extra = try legacyExtra(database: database, position: position),
                   let layer1 = extra.storage(subChunkY: logicalY) {
                    decoded = decoded.withLegacyExtraLayer(layer1)
                }
                return BedrockStoredSubChunk(yIndex: logicalY, subChunk: decoded, backing: .subChunk(key: key))
            }
        }

        if (0...7).contains(Int(yIndex)) {
            let legacyKey = BedrockDBKey(position: position, recordType: .legacyTerrain, subChunkIndex: nil).encoded()
            if let raw = try database.get(legacyKey) {
                let terrain = try BedrockLegacyTerrain.decode(raw)
                return BedrockStoredSubChunk(
                    yIndex: yIndex,
                    subChunk: try terrain.subChunk(yIndex: yIndex),
                    backing: .legacyTerrain(key: legacyKey)
                )
            }
        }

        // Covers extra-only slices and the rare compatibility case where early
        // v9 migrations retain a historical key suffix while carrying absolute
        // Y inside the value.
        return try records(database: database, position: position).first(where: { $0.yIndex == yIndex })
    }

    /// Produces persistent writes for coordinate-keyed edited SubChunks.
    /// - LegacyTerrain slices are merged back into their one 0x30 value.
    /// - Numeric v0/v2...v7 storage 0 is written to 0x2F and virtual storage 1
    ///   is written to the chunk-wide 0x34 LegacyBlockExtraData value.
    /// - Modern storages are written to their normal 0x2F value.
    static func persistentPuts(
        database: MojangLevelDB,
        position: ChunkPosition,
        edited: [Int8: BedrockSubChunk],
        preferLegacyTerrainIfMissing: Bool = false
    ) throws -> [(key: Data, value: Data)] {
        guard !edited.isEmpty else { return [] }
        let existing = try records(database: database, position: position)
        var byY = [Int8: BedrockStoredSubChunk]()
        for record in existing where byY[record.yIndex] == nil { byY[record.yIndex] = record }

        let legacyKey = BedrockDBKey(position: position, recordType: .legacyTerrain, subChunkIndex: nil).encoded()
        let extraKey = BedrockDBKey(position: position, recordType: .legacyBlockExtraData, subChunkIndex: nil).encoded()
        let hasLegacyTerrainBacking = existing.contains { record in
            if case .legacyTerrain = record.backing { return true }
            return false
        }

        if hasLegacyTerrainBacking {
            for y in edited.keys where !(0...7).contains(Int(y)) {
                throw MCBEEditorError.unsupported("LegacyTerrain 世界只保存 Y=0…127；不能直接创建 SubChunk Y=\(y)")
            }
        }

        if existing.isEmpty && preferLegacyTerrainIfMissing {
            var terrain = BedrockLegacyTerrain.empty()
            for y in edited.keys.sorted() {
                guard (0...7).contains(Int(y)), let subChunk = edited[y] else {
                    throw MCBEEditorError.unsupported("LegacyTerrain 世界只保存 Y=0…127")
                }
                try terrain.replaceSubChunk(yIndex: y, with: subChunk)
            }
            return [(legacyKey, try terrain.encodePersistent())]
        }

        var legacyTerrain: BedrockLegacyTerrain?
        if edited.keys.contains(where: {
            if case .legacyTerrain? = byY[$0]?.backing { return true }
            return false
        }), let raw = try database.get(legacyKey) {
            legacyTerrain = try BedrockLegacyTerrain.decode(raw)
        }

        let existingExtraRaw = try database.get(extraKey)
        var extraData = try existingExtraRaw.map(BedrockLegacyBlockExtraData.decode) ?? .empty
        var touchedExtra = false
        var puts = [(key: Data, value: Data)]()
        var touchedLegacyTerrain = false

        for y in edited.keys.sorted() {
            guard let subChunk = edited[y] else { continue }
            if let record = byY[y] {
                switch record.backing {
                case .subChunk(let key):
                    if subChunk.isLegacyNumeric {
                        guard (0...15).contains(Int(y)) else {
                            throw MCBEEditorError.unsupported("LegacyBlockExtraData 只能表示 Y=0…255")
                        }
                        let physical = subChunk.legacyPrimaryLayerOnly()
                        puts.append((key, try physical.encodePersistent()))
                        let layer1 = subChunk.storages.count > 1 ? subChunk.storages[1] : nil
                        try extraData.replaceStorage(subChunkY: y, with: layer1)
                        touchedExtra = true
                    } else {
                        puts.append((key, try subChunk.encodePersistent()))
                        // If this slice used to be numeric, its old 0x34 layer
                        // has either been migrated into the modern storages or
                        // intentionally cleared. Do not leave stale data behind.
                        if record.subChunk.isLegacyNumeric, (0...15).contains(Int(y)) {
                            try extraData.replaceStorage(subChunkY: y, with: nil)
                            touchedExtra = true
                        }
                    }
                case .legacyTerrain:
                    guard var terrain = legacyTerrain else {
                        throw MCBEEditorError.malformedData("LegacyTerrain 写回时原记录不存在")
                    }
                    try terrain.replaceSubChunk(yIndex: y, with: subChunk)
                    legacyTerrain = terrain
                    touchedLegacyTerrain = true
                }
            } else {
                let key = BedrockDBKey.subChunk(
                    x: position.x, z: position.z, dimension: position.dimension, index: y
                )
                if subChunk.isLegacyNumeric {
                    let physical = subChunk.legacyPrimaryLayerOnly()
                    puts.append((key, try physical.encodePersistent()))
                    let layer1 = subChunk.storages.count > 1 ? subChunk.storages[1] : nil
                    try extraData.replaceStorage(subChunkY: y, with: layer1)
                    touchedExtra = true
                } else {
                    puts.append((key, try subChunk.encodePersistent()))
                }
            }
        }
        if touchedLegacyTerrain, let terrain = legacyTerrain {
            puts.append((legacyKey, try terrain.encodePersistent()))
        }
        if touchedExtra {
            // persistentPuts returns puts only. Keep a valid zero-count 0x34
            // record when the last entry is removed rather than requiring an
            // out-of-band delete; whole-chunk modernisation explicitly deletes
            // the legacy key.
            puts.append((extraKey, try extraData.encodePersistent()))
        }
        return puts
    }

    private static func legacyExtra(
        database: MojangLevelDB, position: ChunkPosition
    ) throws -> BedrockLegacyBlockExtraData? {
        let key = BedrockDBKey(
            position: position, recordType: .legacyBlockExtraData, subChunkIndex: nil
        ).encoded()
        guard let raw = try database.get(key) else { return nil }
        return try BedrockLegacyBlockExtraData.decode(raw)
    }
}

private extension BedrockSubChunk {
    func withLegacyExtraLayer(_ layer1: SubChunkStorage) -> BedrockSubChunk {
        guard isLegacyNumeric else { return self }
        var updated = storages
        if updated.isEmpty {
            let air = BedrockBlockState(nbt: nil, legacyID: 0, legacyData: 0)
            updated = [.airFilled(with: air)]
        }
        if updated.count == 1 { updated.append(layer1) }
        else { updated[1] = layer1 }
        return BedrockSubChunk(
            version: version, yIndex: yIndex, storages: updated,
            trailingData: trailingData, rawPersistentData: rawPersistentData
        )
    }

    func legacyPrimaryLayerOnly() -> BedrockSubChunk {
        guard isLegacyNumeric else { return self }
        let primary: [SubChunkStorage] = storages.first.map { [$0] } ?? []
        return BedrockSubChunk(
            version: version, yIndex: yIndex, storages: primary,
            trailingData: trailingData, rawPersistentData: rawPersistentData
        )
    }
}

File: Windows/MCBEEditor.Core/Chunk/BedrockChunkSubChunkAccess.cs
using MCBEEditor.Core.World;

namespace MCBEEditor.Core.Chunk;

public enum BedrockSubChunkBackingKind
{
    SubChunk,
    LegacyTerrain
}

public sealed record BedrockStoredSubChunk(
    sbyte YIndex,
    BedrockSubChunk SubChunk,
    BedrockSubChunkBackingKind BackingKind,
    byte[] BackingKey);

/// <summary>
/// Coordinate-oriented access to normal 0x2F SubChunks, numeric 0x34 extra
/// layer storage and pre-Anvil 0x30 LegacyTerrain columns.
/// </summary>
public sealed class BedrockChunkSubChunkAccess
{
    private readonly IWorldDatabase _database;
    public BedrockChunkSubChunkAccess(IWorldDatabase database) => _database = database;

    public IReadOnlyList<BedrockStoredSubChunk> Records(ChunkPosition position)
    {
        var output = new List<BedrockStoredSubChunk>();
        var occupied = new HashSet<sbyte>();
        (byte[] Key, byte[] Value)? legacyTerrain = null;
        BedrockLegacyBlockExtraData? extra = null;
        byte? observedNumericVersion = null;

        foreach (var entry in _database.Entries(BedrockChunkStore.CoordinatePrefix(position.X, position.Z), includeValues: true))
        {
            if (entry.Value is null || !BedrockDbKey.TryParse(entry.Key, out var key) || key.Position != position) continue;
            if (key.RecordType == ChunkRecordType.SubChunk && key.SubChunkIndex is sbyte keyY)
            {
                var decoded = BedrockSubChunk.Decode(entry.Value, keyY);
                var logicalY = decoded.YIndex ?? keyY;
                if (decoded.IsLegacyNumeric) observedNumericVersion ??= decoded.Version;
                output.Add(new BedrockStoredSubChunk(logicalY, decoded, BedrockSubChunkBackingKind.SubChunk, entry.Key.ToArray()));
                occupied.Add(logicalY);
            }
            else if (key.RecordType == ChunkRecordType.LegacyTerrain)
            {
                legacyTerrain = (entry.Key.ToArray(), entry.Value.ToArray());
            }
            else if (key.RecordType == ChunkRecordType.LegacyBlockExtraData)
            {
                extra = BedrockLegacyBlockExtraData.Decode(entry.Value);
            }
        }

        if (extra is not null)
        {
            for (var index = 0; index < output.Count; index++)
            {
                var record = output[index];
                if (!record.SubChunk.IsLegacyNumeric || record.BackingKind != BedrockSubChunkBackingKind.SubChunk) continue;
                if (extra.Storage(record.YIndex) is not { } layer1) continue;
                var storages = record.SubChunk.Storages.ToList();
                if (storages.Count == 0) storages.Add(SubChunkStorage.AirFilled(new BedrockBlockState(null, 0, 0)));
                if (storages.Count == 1) storages.Add(layer1); else storages[1] = layer1;
                output[index] = record with { SubChunk = record.SubChunk with { Storages = storages } };
            }

            var fallbackVersion = observedNumericVersion ?? 0;
            foreach (var y in extra.Entries.Select(entry => entry.SubChunkY).Distinct().OrderBy(value => value))
            {
                if (occupied.Contains(y) || y is < 0 or > 15 || extra.Storage(y) is not { } layer1) continue;
                var baseSub = BedrockSubChunk.EmptyLegacy(fallbackVersion, y);
                var storages = baseSub.Storages.ToList();
                storages.Add(layer1);
                var key = BedrockDbKey.SubChunk(position.X, position.Z, position.Dimension, y);
                output.Add(new BedrockStoredSubChunk(y, baseSub with { Storages = storages }, BedrockSubChunkBackingKind.SubChunk, key));
                occupied.Add(y);
            }
        }

        if (legacyTerrain is { } terrainRecord)
        {
            var terrain = BedrockLegacyTerrain.Decode(terrainRecord.Value);
            for (sbyte y = 0; y < 8; y++)
            {
                if (occupied.Contains(y)) continue;
                output.Add(new BedrockStoredSubChunk(y, terrain.SubChunk(y), BedrockSubChunkBackingKind.LegacyTerrain, terrainRecord.Key));
            }
        }

        return output.OrderBy(record => record.YIndex)
            .ThenBy(record => record.BackingKind == BedrockSubChunkBackingKind.SubChunk ? 0 : 1)
            .ToArray();
    }

    public BedrockStoredSubChunk? Record(ChunkPosition position, sbyte yIndex)
    {
        var directKey = BedrockDbKey.SubChunk(position.X, position.Z, position.Dimension, yIndex);
        if (_database.Get(directKey) is { } directRaw)
        {
            var decoded = BedrockSubChunk.Decode(directRaw, yIndex);
            var logicalY = decoded.YIndex ?? yIndex;
            if (logicalY == yIndex)
            {
                if (decoded.IsLegacyNumeric && LegacyExtra(position) is { } extra && extra.Storage(logicalY) is { } layer1)
                {
                    var storages = decoded.Storages.ToList();
                    if (storages.Count == 0) storages.Add(SubChunkStorage.AirFilled(new BedrockBlockState(null, 0, 0)));
                    if (storages.Count == 1) storages.Add(layer1); else storages[1] = layer1;
                    decoded = decoded with { Storages = storages };
                }
                return new BedrockStoredSubChunk(logicalY, decoded, BedrockSubChunkBackingKind.SubChunk, directKey);
            }
        }

        if (yIndex is >= 0 and <= 7)
        {
            var legacyKey = new BedrockDbKey(position, ChunkRecordType.LegacyTerrain, null).Encode();
            if (_database.Get(legacyKey) is { } raw)
                return new BedrockStoredSubChunk(yIndex, BedrockLegacyTerrain.Decode(raw).SubChunk(yIndex), BedrockSubChunkBackingKind.LegacyTerrain, legacyKey);
        }

        return Records(position).FirstOrDefault(record => record.YIndex == yIndex);
    }

    public IReadOnlyList<WorldDatabasePut> PersistentPuts(
        ChunkPosition position,
        IReadOnlyDictionary<sbyte, BedrockSubChunk> edited,
        bool preferLegacyTerrainIfMissing = false,
        BedrockEmptyChunkProfile? metadataProfile = null)
    {
        if (edited.Count == 0) return [];
        var existing = Records(position);
        var byY = existing.GroupBy(record => record.YIndex).ToDictionary(group => group.Key, group => group.First());
        var legacyKey = new BedrockDbKey(position, ChunkRecordType.LegacyTerrain, null).Encode();
        var extraKey = new BedrockDbKey(position, ChunkRecordType.LegacyBlockExtraData, null).Encode();
        var hasLegacyTerrain = existing.Any(record => record.BackingKind == BedrockSubChunkBackingKind.LegacyTerrain);
        if (hasLegacyTerrain && edited.Keys.Any(y => y is < 0 or > 7))
            throw new NotSupportedException("LegacyTerrain 世界只保存 Y=0…127；不能直接创建其它 SubChunk。");

        if (existing.Count == 0 && preferLegacyTerrainIfMissing)
        {
            var raw = BedrockLegacyTerrain.EmptyPersistentData;
            var terrain = BedrockLegacyTerrain.Decode(raw);
            foreach (var pair in edited.OrderBy(pair => pair.Key)) ReplaceLegacyTerrainSubChunk(terrain, pair.Key, pair.Value);
            return [new WorldDatabasePut(legacyKey, terrain.EncodePersistent())];
        }

        BedrockLegacyTerrain? legacyTerrain = null;
        if (edited.Keys.Any(y => byY.TryGetValue(y, out var record) && record.BackingKind == BedrockSubChunkBackingKind.LegacyTerrain)
            && _database.Get(legacyKey) is { } legacyRaw)
            legacyTerrain = BedrockLegacyTerrain.Decode(legacyRaw);

        var extraRaw = _database.Get(extraKey);
        var extra = extraRaw is null ? new BedrockLegacyBlockExtraData() : BedrockLegacyBlockExtraData.Decode(extraRaw);
        var touchedExtra = false;
        var touchedLegacyTerrain = false;
        var puts = new List<WorldDatabasePut>();

        foreach (var pair in edited.OrderBy(pair => pair.Key))
        {
            var y = pair.Key;
            var subChunk = pair.Value;
            if (!subChunk.IsLegacyNumeric && !subChunk.IsRawPreservedUnknownVersion)
            {
                var paletteFormat = byY.TryGetValue(y, out var original)
                    ? BedrockPaletteFormat.Detect(original.SubChunk.Storages.SelectMany(storage => storage.Palette))
                    : BedrockPaletteFormat.Detect(existing.SelectMany(record => record.SubChunk.Storages).SelectMany(storage => storage.Palette))
                        ?? BedrockPaletteFormat.Detect(subChunk.Storages.SelectMany(storage => storage.Palette));
                if (paletteFormat?.UsesLegacyVal == true)
                    subChunk = subChunk with { Storages = subChunk.Storages.Select(storage => storage with
                    {
                        Palette = storage.Palette.Select(state => BedrockLegacyBlockStateConverter.ForPalette(state, paletteFormat)).ToArray()
                    }).ToArray() };
            }
            if (byY.TryGetValue(y, out var record))
            {
                if (record.BackingKind == BedrockSubChunkBackingKind.LegacyTerrain)
                {
                    legacyTerrain ??= _database.Get(legacyKey) is { } raw ? BedrockLegacyTerrain.Decode(raw) : throw new InvalidDataException("LegacyTerrain 写回时原记录不存在。");
                    ReplaceLegacyTerrainSubChunk(legacyTerrain, y, subChunk);
                    touchedLegacyTerrain = true;
                }
                else if (subChunk.IsLegacyNumeric)
                {
                    var primary = subChunk with { Storages = subChunk.Storages.Take(1).ToArray() };
                    puts.Add(new WorldDatabasePut(record.BackingKey.ToArray(), primary.EncodePersistent()));
                    extra.ReplaceStorage(y, subChunk.Storages.Count > 1 ? subChunk.Storages[1] : null);
                    touchedExtra = true;
                }
                else
                {
                    puts.Add(new WorldDatabasePut(record.BackingKey.ToArray(), subChunk.EncodePersistent()));
                    if (record.SubChunk.IsLegacyNumeric && y is >= 0 and <= 15)
                    {
                        extra.ReplaceStorage(y, null);
                        touchedExtra = true;
                    }
                }
            }
            else
            {
                var key = BedrockDbKey.SubChunk(position.X, position.Z, position.Dimension, y);
                if (subChunk.IsLegacyNumeric)
                {
                    var primary = subChunk with { Storages = subChunk.Storages.Take(1).ToArray() };
                    puts.Add(new WorldDatabasePut(key, primary.EncodePersistent()));
                    extra.ReplaceStorage(y, subChunk.Storages.Count > 1 ? subChunk.Storages[1] : null);
                    touchedExtra = true;
                }
                else
                {
                    puts.Add(new WorldDatabasePut(key, subChunk.EncodePersistent()));
                }
            }
        }

        if (touchedLegacyTerrain && legacyTerrain is not null)
            puts.Add(new WorldDatabasePut(legacyKey, legacyTerrain.EncodePersistent()));
        if (touchedExtra)
            puts.Add(new WorldDatabasePut(extraKey, extra.EncodePersistent()));
        if (!hasLegacyTerrain)
            puts.AddRange(BedrockEmptyChunkMetadata.MissingRecords(_database, position, metadataProfile));
        return puts;
    }

    private BedrockLegacyBlockExtraData? LegacyExtra(ChunkPosition position)
    {
        var key = new BedrockDbKey(position, ChunkRecordType.LegacyBlockExtraData, null).Encode();
        return _database.Get(key) is { } raw ? BedrockLegacyBlockExtraData.Decode(raw) : null;
    }

    private static void ReplaceLegacyTerrainSubChunk(BedrockLegacyTerrain terrain, sbyte yIndex, BedrockSubChunk subChunk)
    {
        if (yIndex is < 0 or > 7) throw new NotSupportedException("LegacyTerrain 只包含 SubChunk Y=0…7。");
        if (!subChunk.IsLegacyNumeric || subChunk.Storages.Count != 1)
            throw new NotSupportedException("LegacyTerrain 只能写入单层旧版数字 ID SubChunk。");
        var storage = subChunk.Storages[0];
        if (storage.Indices.Length != 4096 || storage.Palette.Count == 0)
            throw new InvalidDataException("LegacyTerrain 替换 storage 无效。");
        for (var x = 0; x < 16; x++)
        for (var z = 0; z < 16; z++)
        for (var localY = 0; localY < 16; localY++)
        {
            var blockIndex = (x << 8) | (z << 4) | localY;
            var paletteIndex = storage.Indices[blockIndex];
            if (paletteIndex >= storage.Palette.Count) throw new InvalidDataException("LegacyTerrain 调色板索引越界。");
            var state = storage.Palette[paletteIndex];
            if (state.Nbt is not null || state.LegacyId is not ushort legacyId || legacyId > byte.MaxValue)
                throw new NotSupportedException("LegacyTerrain 只能写入 0…255 数字 ID。");
            var absoluteY = yIndex * 16 + localY;
            var sourceIndex = (x << 11) | (z << 7) | absoluteY;
            terrain.BlockIds[sourceIndex] = (byte)legacyId;
            var packedIndex = sourceIndex >> 1;
            var data = (byte)((state.LegacyData ?? 0) & 0x0f);
            terrain.DataValues[packedIndex] = (sourceIndex & 1) == 0
                ? (byte)((terrain.DataValues[packedIndex] & 0xf0) | data)
                : (byte)((terrain.DataValues[packedIndex] & 0x0f) | (data << 4));
        }
    }
}
